### PR TITLE
Project recently accessed (Top 5)

### PR DESCRIPTION
**Pull request by bhatiadheeraj**

### CODE REVIEW

File: ui/src/projects.vue [hunk 1]
```diff
@@ -180,6 +196,12 @@ export default {
                     } else {
                         this.other_projects.push(p);
                     }
+                    if(localStorage.getItem("recent_projectid")){
+                        const recentProjects = JSON.parse(localStorage.getItem("recent_projectid")).slice(0).slice(-5);
+                        if(recentProjects && recentProjects.includes(p._id.toString())) {
+                            this.recentProjects.push(p);
```
Comment (soichih) [inline ui/src/projects.vue:202]: Won't the list keep growing? Also.. if I visit a group, then  visit other groups (at least 5 of them), then the first group will never show on the recent projects list again, as that group is already listed in the array to the left?

**Review by soichih — CHANGES_REQUESTED** [review]

File: ui/src/projects.vue (hunk 1 above)
Comment (bhatiadheeraj) [reply ui/src/projects.vue:202]: Yes, the list would keep growing and it's listed to the left!

File: ui/src/projects.vue [hunk 2]
```diff
@@ -24,7 +24,22 @@
         <div v-if="query.length && !other_projects.length && !my_projects.length">
             <p style="padding: 20px; opacity: 0.8;">No matching Projects</p>
         </div>
-
+        <div v-if="config.user && recentProjects.length" class="position: relative">
```
[the remark this answers is not in the document]
Comment (bhatiadheeraj) [reply ui/src/projects.vue:27]: yes, we should!

File: ui/src/projects.vue (hunk 1 above)
Comment (soichih) [reply ui/src/projects.vue:202]: Will you be able to fix it?

Comment (bhatiadheeraj) [reply ui/src/projects.vue:202]: the project will show if it is accessed again. If not then only the last 5 will show up. I will still make always have no more than 5 projects.

File: ui/src/projects.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (bhatiadheeraj) [reply ui/src/projects.vue:27]: how should we display this component?

**Comment by soichih** [comment]
One feature that I've always wished we had was to show project that a user recently became a member. Like.. when someone add you to a member, right now we don't really have a good way for the user to know which project that the user was added.

We could show both recently accessed project *and* projects that user was recently added as a member - as well as any project that user recently created.  I believe we can implement this by doing the following.

1) for every project that user visits, store the timestamp on local storage.
2) while iterating over project list, look for any project that user does not have the timestamp (added recently) and set it to the current timestamp.
3) sort the timestamps/project list descending order. pick the first 5 and add those to the "recent project" list.

What do you think?

**Comment by soichih** [comment]
@bhatiadheeraj The PR doesn't look like it shows the recently added project as "recent" project. It only shows a project that the user has personally accessed most recently. Are you still working on that? Or should I go ahead and merge this and maybe I should work on it?

**Comment by bhatiadheeraj** [comment]
@soichih  I am working on any project the user does not have the timestamp (added recently) and set it to the current timestamp.

**Comment by soichih** [comment]
Amending on the algorithm..

1. for every project that a user visits, store the timestamp on local storage.
2. when the project list is loaded from the API, look for any project that does not have the timestamp on the local storage (it means user hasn't visited the project since we added this feature - or it could be a new project that someone shared with the user). Use *current* timestamp for such project and show "New" sign on the UI (until user visit that project).
3. sort the timestamps/project list in descending order. pick the first 5 and add those to the "recent project" list.

Initially, all users will see all 5 projects with "new" signs incorrectly - as we can't distinguish between projects that user hasn't visited yet, or was recently added. I believe we can deal with this by storing both the last accessed time and last *seen* data returned from the API.. but I need to do a bit more thinking on this.

File: ui/src/projects.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (soichih) [reply ui/src/projects.vue:27]: @bhatiadheeraj You saw me yesterday how I refactored this to <projectblock> component, right? Are you able to refactor this? (and probably set some appropriate stylesheet for "New" label? I think something like this should work..

image

```
<b-badge variant="primary">New</b-badge>
```

File: ui/src/projects.vue [hunk 3]
```diff
@@ -21,10 +21,26 @@
...
 
-            <div v-if="query.length && !other_projects.length && !my_projects.length">
-                <p style="padding: 20px; opacity: 0.8;">No matching Projects</p>
+        <div v-if="query.length && !other_projects.length && !my_projects.length">
+            <p style="padding: 20px; opacity: 0.8;">No matching Projects</p>
+        </div>
+        <div v-if="config.user && recentProjects.length" class="position: relative">
+            <h4 class="group-title">Recent Projects</h4>
+            <div style="padding: 10px;" v-if="mode == 'tile'">
+                <div v-for="project in recentProjects.slice(0).slice(-5)" :key="project._id">
```
Comment (soichih) [inline ui/src/projects.vue:30]: I think we can just show *all* recent projects - instead of top 5.

**Review by soichih — CHANGES_REQUESTED** [review]

**Pull request merged**